c7n/actions.py

```diff
@@ -186,26 +186,115 @@ class ModifyVpcSecurityGroupsAction(Action):
+        """
+        if r.get('Groups'):
+            if metadata_key and isinstance(r['Groups'][0], dict):
+                rgroups = [g[metadata_key] for g in r['SecurityGroups']]
+            else:
+                rgroups = [g['GroupId'] for g in r['Groups']]
+        elif r.get('SecurityGroups'):
+            # elb, ec2, elasticache, efs, dax vpc resource security groups
+            if metadata_key and isinstance(r['SecurityGroups'][0], dict):
+                rgroups = [g[metadata_key] for g in r['SecurityGroups']]
+            else:
+                rgroups = [g for g in r['SecurityGroups']]
+        elif r.get('VpcSecurityGroups'):
+            # rds resource security groups
+            if metadata_key and isinstance(r['VpcSecurityGroups'][0], dict):
+                rgroups = [g[metadata_key] for g in r['VpcSecurityGroups']]
+            else:
+                rgroups = [g for g in r['VpcSecurityGroups']]
+        elif r.get('VPCOptions', {}).get('SecurityGroupIds', []):
+            # elasticsearch resource security groups
+            if metadata_key and isinstance(
+                    r['VPCOptions']['SecurityGroupIds'][0], dict):
+                rgroups = [g[metadata_key] for g in r[
+                    'VPCOptions']['SecurityGroupIds']]
+            else:
+                rgroups = [g for g in r['VPCOptions']['SecurityGroupIds']]
+        # use as substitution for 'Groups' or '[Vpc]SecurityGroups'
+        # unsure if necessary - defer to coverage report
+        elif metadata_key and r.get(metadata_key):
+            rgroups = [g for g in r[metadata_key]]
+
+        return rgroups
+
+    def parse_groups(self, r, target_group_ids, rgroups, action):
+        """
+        Parse user-provided groups in policy and resolves security groups
+        from either names or whitelisted names (matched, network-location, all)
+        """
+        groups = []
+        names = []
+
+        if action == 'remove':
+            # Parse remove_groups
+            if target_group_ids == 'matched':
+                return r.get('c7n:matched-security-groups', ())
+            elif target_group_ids == 'network-location':
+                for reason in r.get('c7n:NetworkLocation', ()):
+                    if reason['reason'] == 'SecurityGroupMismatch':
+                        groups = list(reason['security-groups'])
+                return groups
+            elif target_group_ids == 'all':
+                return rgroups
+
+        if isinstance(target_group_ids, list):
+            groups = [sg_id for sg_id in target_group_ids if sg_id.startswith('sg-')]
+            names = [name for name in target_group_ids if not name.startswith('sg-')]
+        elif isinstance(target_group_ids, six.string_types):
+            if target_group_ids.startswith('sg-'):
+                groups = [target_group_ids]
+            else:
+                names = [target_group_ids]
+
+        if names:
+            groups += self.resolve_security_group_names(names)
```

Comment:
doing the resolution here means we do an api call per resource to resolve names instead of doing it one time for the whole set of resources.

Reply:
oops don't want to do that